src/core/orderbook.js

```diff
@@ -933,41 +933,44 @@ OrderBook.prototype.insertOffer = function(node) {
     log.info('inserting offer', this._key, node.fields);
   }
 
-  const offer = OrderBook.offerRewrite(node.fields);
-  const takerGets = this.normalizeAmount(this._currencyGets, offer.TakerGets);
-  const takerPays = this.normalizeAmount(this._currencyPays, offer.TakerPays);
+  try {
+    const offer = OrderBook.offerRewrite(node.fields);
+    const takerGets = this.normalizeAmount(this._currencyGets, offer.TakerGets);
+    const takerPays = this.normalizeAmount(this._currencyPays, offer.TakerPays);
 
-  offer.LedgerEntryType = node.entryType;
-  offer.index = node.ledgerIndex;
+    offer.LedgerEntryType = node.entryType;
+    offer.index = node.ledgerIndex;
 
-  // We're safe to calculate quality for newly created offers
-  offer.quality = takerPays.divide(takerGets).to_text();
+    // We're safe to calculate quality for newly created offers
+    offer.quality = takerPays.divide(takerGets).to_text();
 
-  const originalLength = this._offers.length;
+    const originalLength = this._offers.length;
 
-  for (let i = 0; i < originalLength; i++) {
-    const quality = OrderBookUtils.getOfferQuality(offer, this._currencyGets);
-    const existingOfferQuality = OrderBookUtils.getOfferQuality(
-      this._offers[i],
-      this._currencyGets
-    );
+    for (let i = 0; i < originalLength; i++) {
+      const quality = OrderBookUtils.getOfferQuality(offer, this._currencyGets);
+      const existingOfferQuality = OrderBookUtils.getOfferQuality(
+        this._offers[i],
+        this._currencyGets
+      );
 
-    if (quality.compareTo(existingOfferQuality) <= 0) {
-      this._offers.splice(i, 0, offer);
+      if (quality.compareTo(existingOfferQuality) <= 0) {
+        this._offers.splice(i, 0, offer);
 
-      break;
+        break;
+      }
     }
-  }
 
-  if (this._offers.length === originalLength) {
-    this._offers.push(offer);
-  }
-
-  this.incrementOwnerOfferCount(offer.Account);
-
-  this.updateOwnerOffersFundedAmount(offer.Account);
+    if (this._offers.length === originalLength) {
+      this._offers.push(offer);
+    }
 
-  this.emit('offer_added', offer);
+    this.incrementOwnerOfferCount(offer.Account);
+    this.updateOwnerOffersFundedAmount(offer.Account);
+    this.emit('offer_added', offer);
+  } catch (e) {
```

Comment:
I you run reconnect() on a timeout from inside the connect() method of Server, you can fairly easily reproduce the bug without elaborate setup etc. If you do this, another bug will be exposed, in the _sendMessage method, where it tries to send a message when the socket isn't open.

```javascript
Server.prototype._sendMessage = function(message) {
  if (this._ws) {
    if (this._remote.trace) {
```

You need to add something like  `&& this._socket_open` (set true of course when the, err, socket is open ) 

Here's a list of (relatively) active orderbooks to subscribe to:
https://gist.github.com/sublimator/995713d2fcff383e7221

You can `fix` the bug (read: make it much much less likely to happen) with:
```
OrderBook.prototype.resetCache = function() {
  this._subscribed = false;
```

A solid fix would need ensuring we have consistent states before applying transactions, and making sure we apply all transactions in a ledger, in the right order (say we are connected to multiple servers, even though we dedupe transactions by hash before notifying listeners, I don't recall that we ensured any order )